### PR TITLE
simplify/clarify README.md

### PR DESCRIPTION
**Pull request by sole**
I have copywritten this thing and also added a bunch of TODO where things are not clear or confusing

Let me know what you think :P

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -57,24 +73,29 @@ worker.get('/myResource.html', function(request, response) {
 });
 ```
 
-## Why I would like to support more than just handling requests?
+TODO: a working code example that can be executed to see how 'Your code' would actually look like
```
Comment (arcturus) [inline README.md]: We could use the following:
`importScripts('./sww.js');
var worker = new self.ServiceWorkerWare();
worker.get('/hello.html', function(request, response) {
  return Promise.resolve(new Response('Hello world!), { headers: {'Content-Type': 'text/html'} });
}
worker.init();`

File: README.md [hunk 2]
```diff
@@ -57,24 +73,29 @@ worker.get('/myResource.html', function(request, response) {
...
+
+### onInstall(?) TODO
```
Comment (arcturus) [inline README.md]: Will be called during ServiceWorker installation. This happens just once.

File: README.md [hunk 3]
```diff
@@ -57,24 +73,29 @@ worker.get('/myResource.html', function(request, response) {
...
+### onActivate(?) TODO
```
Comment (arcturus) [inline README.md]: Called just after `onInstall` or each time that you update your ServiceWorker, useful to update caches or any part of your application logic.

File: README.md [hunk 4]
```diff
@@ -57,24 +73,29 @@ worker.get('/myResource.html', function(request, response) {
...
 
-## Can I use more than one middleware combined?
+* `request`: it's a standard [Request Object](https://fetch.spec.whatwg.org/#concept-request)
+* `response`: it's a standard [Response Object](https://fetch.spec.whatwg.org/#concept-response)
 
-Right, you can do it, take into account that the request/response objects will be passing through them sequentially in the order that we register our middleware objetcs.
+You need to `.clone()` the `request` before using it as some fields are one use only.
 
-## Handling requests
+TODO: What is the use of the `response` parameter?
```
Comment (arcturus) [inline README.md]: After the whole process of iterating over the different middleware you need to return a Response object, different middleware could use the previous response object to perform any operations over content, headers or anything.

File: README.md [hunk 5]
```diff
@@ -90,43 +111,73 @@ worker.get('virtual/.', function(request, response) {
 });
 ```
 
-The output for any request that hits this url will be the original address of the url and a random number, with the headers that we specified in the response object.
+The output for any request that hits this URL will be the original address of the URL, and a randomly generated number. Note how we can even specify the headers for our response!
+
+Remember that the handler *must always return a Promise*.
+
+### onMessage(?) TODO
```
Comment (arcturus) [inline README.md]: Will receive this event if an actor (window, iframe, another worker or sharedworker) perform a `postMessage` to the ServiceWorker to communicate with it.

File: README.md [hunk 6]
```diff
@@ -90,43 +111,73 @@ worker.get('virtual/.', function(request, response) {
...
+### onBeforeEvicted(?) TODO
```
Comment (arcturus) [inline README.md]: (Still under heavy discussion): It's a custom event that the UA send to the worker to let it know that is going to be evicted

File: README.md [hunk 7]
```diff
@@ -90,43 +111,73 @@ worker.get('virtual/.', function(request, response) {
...
+### onEvicted(?) TODO
```
Comment (arcturus) [inline README.md]: I'm thinking on not documenting this two methods, as they are under heavy discussion now, leave them for experiment, not so sure, what do you think?

File: README.md [hunk 8]
```diff
@@ -90,43 +111,73 @@ worker.get('virtual/.', function(request, response) {
...
+
 
-One important point is the fact that when you write a handler for using in your request, the function handling those requests *must return a promise*.
+## Middleware examples
 
-## Can I see some examples of middleware?
+This package already incorporates two simple middlewares. They are available by default when you import it:
 
-Right we added 2 simple middlewares to this package, one for precaching static content and a simple offline cache handler. They are available by default when you import this library.
+TODO: I would suggest refactoring them away from this package
```
Comment (arcturus) [inline README.md]: Sounds good to me, did that with the cache helper.

Will do it during the flight :P

File: README.md [hunk 9]
```diff
@@ -90,43 +111,73 @@ worker.get('virtual/.', function(request, response) {
...
+
+### `StaticCacher`
+
+This will let you preload and cache static content during the `ServiceWorker` installation.
+
+TODO: where does `self` come from? in which context is this being executed?
```
Comment (arcturus) [inline README.md]: `self` is the `this` object in the context of a serviceworker

File: README.md [hunk 10]
```diff
@@ -90,43 +111,73 @@ worker.get('virtual/.', function(request, response) {
...
 
-* SimpleOfflineCache: will serve the contents of the default cache. Right now if it cannot find an element in the cache will perform a fetch and will save it to the cache.
-(TODO: this should be configurable ;P)
+Upon installation, this worker will load `a.html` and `b.html` and store their content in the default cache.
+
+### `SimpleOfflineCache`
+
+This will serve contents stored in the default cache. If it cannot find a resource in the cache, it will perform a fetch and save it to the cache.
+
+TODO: this should be configurable ;P // TODO this comment should be in the code for the middleware
```
Comment (arcturus) [inline README.md]: Right :)

File: README.md [hunk 11]
```diff
@@ -90,43 +111,73 @@ worker.get('virtual/.', function(request, response) {
...
+
+### [<tt>ZipCacher</tt>](https://github.com/arcturus/zipcacher)
 
-### More examples of middleware
-* [ZipCacher](https://github.com/arcturus/zipcacher) enables you to specify a zip file to cache your resources from.
+This is not built-in with this library. It enables you to specify a ZIP file to cache your resources from.
 
-# Demo
+TODO: example
```
Comment (arcturus) [inline README.md]:
```
importScripts('./sww.js');
importScripts('./zipcacher.js');

var worker = new self.ServiceWorkerWare();

worker.use(new ZipCacher('http://localhost:8000/demo/resources.zip'));
worker.use(new self.SimpleOfflineCache());
worker.init();

File: README.md [hunk 12]
```diff
@@ -90,43 +111,73 @@ worker.get('virtual/.', function(request, response) {
...
+Clone the repository, and then cd to the directory and run:
 
-Please run:
 ```
 npm install
 gulp webserver
 ```
 
-And go to http://localhost:8000/demo/index.html
+TODO: use npm scripts to run gulp, avoid global gulp installation
```
Comment (arcturus) [inline README.md]: Great point, will you send the PR for that ?

File: README.md [hunk 13]
```diff
@@ -90,43 +111,73 @@ worker.get('virtual/.', function(request, response) {
...
-A lot of this code has been written with the inspiration from different projects:
 - [Firefox OS V3 Architecture] (https://github.com/fxos/contacts)
 - [Shed] (https://github.com/wibblymat/shed)
-- [sw-preache](https://github.com/jeffposnick/sw-precache)
+- [sw-precache](https://github.com/jeffposnick/sw-precache)
 - [offliner](https://github.com/lodr/offliner)
+
+## License
+
+??? TODO ???
```
Comment (arcturus) [inline README.md]: MPL!

File: README.md (hunk 7 above)
Comment (sole) [inline README.md]: I'm commenting them out and adding a note - they'll be still visible in the raw MD file but not when rendered in the project page

File: README.md (hunk 8 above)
Comment (sole) [inline README.md]: Are those https://github.com/arcturus/dynamiccache and https://github.com/arcturus/dynamiccacheclient ?

File: README.md (hunk 9 above)
Comment (sole) [inline README.md]: but if they are refactored to another module the example that uses `self` is invalid, right?

File: README.md (hunk 12 above)
Comment (sole) [inline README.md]: let me add an issue so I don't forget!

**Pull request merged**